test/metabase/driver_test.clj

```diff
@@ -68,24 +72,75 @@
 
+(defn- basic-db-definition [database-name]
+  (tx/map->DatabaseDefinition
+   {:database-name     database-name
+    :table-definitions [{:table-name        "baz"
+                         :field-definitions [{:field-name "foo", :base-type :type/Text}]
+                         :rows              [["bar"]]}]}))
+
+(deftest check-can-connect-before-sync-test
+  (testing "Database sync should short-circuit and fail if the database at the connection has been deleted (metabase#7526)"
+    (mt/test-drivers (->> (mt/normal-drivers)
+                          ;; athena is a special case because connections aren't made with a single database,
+                          ;; but to an S3 bucket that may contain many databases
+                          ;; TODO: other drivers are still failing with this test. For these drivers there's a good chance there's a bug in
+                          ;; test.data.<driver> code, and not with the driver itself.
+                          (remove #{:athena :presto-jdbc :oracle :vertica}))
+      (let [database-name (mt/random-name)
+            dbdef         (basic-db-definition database-name)]
+        (mt/dataset dbdef
+          (let [db (mt/db)
+                cant-sync-logged? (fn []
+                                 (some?
+                                  (some
+                                   (fn [[log-level throwable message]]
+                                     (and (= log-level :warn)
+                                          (instance? clojure.lang.ExceptionInfo throwable)
+                                          (re-matches #"^Cannot sync Database (.+): (.+)" message)))
+                                   (mt/with-log-messages-for-level :warn
+                                     (#'task.sync-databases/sync-and-analyze-database*! (u/the-id db))))))]
+            (testing "sense checks before deleting the database"
+              (testing "sense check 1: sync-and-analyze-database! should not log a warning"
+                (is (false? (cant-sync-logged?))))
+              (testing "sense check 2: triggering the sync via the POST /api/database/:id/sync_schema endpoint should succeed"
+                (is (= {:status "ok"}
+                       (mt/user-http-request :crowberto :post 200 (str "/database/" (u/the-id db) "/sync_schema"))))))
+            ;; release db resources like connection pools we don't have to wait to finish syncing before destroying the db
+            (driver/notify-database-updated driver/*driver* db)
+            ;; destroy the db
+            (if (contains? #{:redshift :snowflake} driver/*driver*)
+              ;; in the case of some cloud databases, the test database is never created, and shouldn't be destroyed.
+              ;; so fake it by redefining the database name on the dbdef
+              (t2/update! :model/Database (u/the-id db)
+                          {:details (assoc (:details (mt/db)) :db "fake-db-name-that-definitely-wont-be-used")})
```

Comment:
I think we can just do this update for all drivers, it'll make connecting to the DB fail anw.

Reply:
The problem with that is this code depends on the key(s) in `details` that identifies the database, and the specifics of how the driver is tested. So we'd really prefer not to have to do it this way, and I wanted to avoid creating another driver multimethod for `metabase.test.data.interface` just for this test.

For example with redshift, we always use a single test DB, identified by the `:db` key ([here](https://github.com/metabase/metabase/blob/6360f1a5f6be43abfdcff6da98e9aac1c8416381/modules/drivers/redshift/test/metabase/test/data/redshift.clj#L45)). So to test what happens when a database is deleted, the test needs to know that `:db` is the key that identifies a DB for redshift.